### PR TITLE
Polymorphic numeric getters for ColumnValueSelector

### PR DESCRIPTION
**Pull request by leventov**
Add methods getFloat(), getDouble() and getLong() to ColumnValueSelector. Then, `LongColumnSelector` has default implementations for `getFloat()` and `getDouble()` which convert the result of `getLong()`. Correspondingly in `FloatColumnSelector` and `DoubleColumnSelector`, with different methods left abstract. `ObjectColumnSelector` has default implementations for all three methods, which cast to `Number`, and `DimensionSelector` simply throws `UnsupportedOperationException`.

This PR is a first step for #4622. It is needed, because currently rollup aggregations could process any numbers in metrics, because they use constructions like `((Number) metric).floatValue()`, when metric value is either `Long`, `Float` or `Double`. With the new ColumnValueSelector-based approach (see #4622), ColumnValueSelector should have similar polymorphism capability.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/LongColumnSelector.java [hunk 1]
```diff
@@ -27,5 +27,32 @@
 public interface LongColumnSelector extends ColumnValueSelector, HotLoopCallee
 {
   @CalledFromHotLoop
-  public long get();
+  @Override
+  long getLong();
+
+  /**
+   * @deprecated This method is marked as deprecated in FloatColumnSelector to minimize the probability of accidential
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/LongColumnSelector.java]: do you mean `in LongColumnSelector` instead of `in FloatColumnSelector` ?

Comment (leventov) [inline processing/src/main/java/io/druid/segment/LongColumnSelector.java]: Thanks, copy-paste mistake

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 2]
```diff
@@ -117,4 +117,34 @@
    */
   @Nullable
   IdLookup idLookup();
+
+  /**
+   * @deprecated always throws {@link UnsupportedOperationException}
+   */
+  @Deprecated
+  @Override
+  default float getFloat()
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: if this is suppose to be always the case why not making this as final methods instead of defaults ?

File: processing/src/main/java/io/druid/segment/ObjectColumnSelector.java [hunk 3]
```diff
@@ -19,12 +19,50 @@
 
 package io.druid.segment;
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/ObjectColumnSelector.java]: stopped here will finish this later.

File: processing/src/main/java/io/druid/segment/DimensionSelector.java (hunk 2 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: `default` interface methods cannot be `final`

File: processing/src/main/java/io/druid/segment/ColumnValueSelector.java [hunk 4]
```diff
@@ -21,7 +21,16 @@
...
 /**
  * Base type for interfaces that manage column value selection, e.g. DimensionSelector, LongColumnSelector
+ *
+ * This interface has methods to get the value in all primitive types, that have corresponding basic aggregators in
+ * Druid: Sum, Min, Max, etc: {@link #getFloat()}, {@link #getDouble()} and {@link #getLong()} to support "polymorphic"
+ * rollup aggregation during index merging.
  */
 public interface ColumnValueSelector
 {
+  float getFloat();
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/ColumnValueSelector.java]: What do you think about adding a new interface maybe ```PrimitiveColumnSelector``` which extends ```ColumnValueSelector``` and is extended by ```Long/Float/DoubleColumnSelector``` and maybe ```ObjectColumnSelector```? If so, ```DimensionSelector``` can extend ```ColumnValueSelector``` directly and don't have to implement unnecessary methods. I'm not sure what the best name is anyway.

File: processing/src/main/java/io/druid/segment/DoubleColumnSelector.java [hunk 5]
```diff
@@ -26,5 +26,32 @@
 public interface DoubleColumnSelector extends ColumnValueSelector, HotLoopCallee
 {
   @CalledFromHotLoop
-  double get();
+  @Override
+  double getDouble();
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/DoubleColumnSelector.java]: What do you think about renaming these methods as ```asDouble()```, ```asFloat()```, and ```asLong()```, and adding another method ```get()``` which calls ```getDouble()``` by default? It's more intuitive to me.
Same for other ColumnValueSelectors.

File: processing/src/main/java/io/druid/segment/ColumnValueSelector.java (hunk 4 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/ColumnValueSelector.java]: I considered that but decided that it's not worthwhile. The interface should probably be called `MetricColumnValueSelector`, because ObjectColumnSelector is not "primitive", but Long/Float/DoubleColumnSelector could also appear as dimension column. Also it's against the general trend into converging "dimensions" and "metrics" into just general "columns" which could be used interchangeably. `DimensionSelector` also could add reasonable implementation for the new methods later.

File: processing/src/main/java/io/druid/segment/DoubleColumnSelector.java (hunk 5 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DoubleColumnSelector.java]: `asDouble()` is "wrong" because by convention it means that that selector object is converted to double. So probably methods should be called `getAsDouble()`, `getAsLong()` and `getAsFloat()`. But it creates inconsistency with `Aggregator` interface. Also "As" method names are longer, and I hardly think anyone could be confused by the current names (e. g. think about them as they have `ByteBuffer.getDouble()` semantics). So I'm slightly pro keeping the current names.

Regarding `get()` overload, I specifically wanted to remove conflict between `Double/Float/LongColumnSelector` interfaces, so that some class could implement all three. It could simplify implementation of some ColumnValueSelector factories. Also, I don't see why having two methods with different names but the same semantics is better than having just one.

File: processing/src/main/java/io/druid/query/groupby/epinephelinae/column/DoubleGroupByColumnSelectorStrategy.java [hunk 6]
```diff
@@ -46,7 +45,19 @@ public void processValueFromGroupingKey(
...
+  public Object getOnlyValue(ColumnValueSelector selector)
+  {
+    return selector.getDouble();
+  }
+
+  @Override
+  public void writeToKeyBuffer(int keyBufferPosition, Object obj, ByteBuffer keyBuffer)
+  {
+    keyBuffer.putDouble(keyBufferPosition, (Double) obj);
   }
```
Comment (xanec) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/column/DoubleGroupByColumnSelectorStrategy.java]: Maybe we can combine `getOnlyValue` and `writeToKeyBuffer` into a single method to avoid additional boxing and unboxing (similarly for `FloatGroupByColumnSelectorStrategy` and `LongGroupByColumnSelectorStrategy`)?

Please correct me if I am wrong but it seems that these two methods are always used together?

Comment (leventov) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/column/DoubleGroupByColumnSelectorStrategy.java]: @xanec It could be a subject for a separate PR

File: processing/src/main/java/io/druid/segment/virtual/ExpressionSelectors.java [hunk 7]
```diff
@@ -105,7 +105,7 @@ public static DoubleColumnSelector makeDoubleColumnSelector(
     class ExpressionDoubleColumnSelector implements DoubleColumnSelector
     {
       @Override
-      public double get()
+      public double getDouble()
       {
         final Double number = baseSelector.get().asDouble();
```
Comment (xanec) [inline processing/src/main/java/io/druid/segment/virtual/ExpressionSelectors.java]: I believe this is out of scope of the PR but I think this poses a potential `NullPointerException`.

Comment (xanec) [inline processing/src/main/java/io/druid/segment/virtual/ExpressionSelectors.java]: #4666

File: processing/src/main/java/io/druid/segment/ColumnValueSelector.java (hunk 4 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/ColumnValueSelector.java]: Sounds make sense.

File: processing/src/main/java/io/druid/segment/DoubleColumnSelector.java (hunk 5 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/DoubleColumnSelector.java]: > Also "As" method names are longer, and I hardly think anyone could be confused by the current names (e. g. think about them as they have ByteBuffer.getDouble() semantics). So I'm slightly pro keeping the current names.

I was confused because I thought ```DoubleColumnSelector``` is specialized for the double type and will have a method to retrieve double values like ```java.nio.DoubleBuffer``` has a series of ```get()``` which return only doubles.

I looked into more codes around index merger and understand why you chose this design. I don't see any better design for now. Please go for it.

File: processing/src/main/java/io/druid/query/groupby/epinephelinae/column/DoubleGroupByColumnSelectorStrategy.java (hunk 6 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/column/DoubleGroupByColumnSelectorStrategy.java]: @xanec, I added them in https://github.com/druid-io/druid/pull/4576, and guess it's possible to combine them. Are you interested in making a patch to fix it?
